linera-views/src/views/historical_hash_wrapper.rs

```diff
@@ -0,0 +1,528 @@
+    /// Track context type.
+    _phantom: PhantomData<C>,
+}
+
+/// Key tags to create the sub-keys of a `HistoricallyHashableView` on top of the base key.
+#[repr(u8)]
+enum KeyTag {
+    /// Prefix for the indices of the view.
+    Inner = MIN_VIEW_TAG,
+    /// Prefix for the hash.
+    Hash,
+}
+
+impl<C, W> HistoricallyHashableView<C, W> {
+    fn make_hash(&self, batch: &Batch) -> Result<HasherOutput, ViewError> {
+        #[cfg(with_metrics)]
+        let _hash_latency = metrics::HISTORICALLY_HASHABLE_VIEW_HASH_RUNTIME.measure_latency();
+        let stored_hash = self.stored_hash.unwrap_or_default();
+        if batch.is_empty() {
+            return Ok(stored_hash);
+        }
+        let mut hasher = sha3::Sha3_256::default();
+        hasher.update_with_bytes(&stored_hash)?;
+        hasher.update_with_bcs_bytes(&batch)?;
+        Ok(hasher.finalize())
+    }
+}
+
+impl<C, W, C2> ReplaceContext<C2> for HistoricallyHashableView<C, W>
+where
+    W: View<Context = C> + ReplaceContext<C2>,
+    C: Context,
+    C2: Context,
+{
+    type Target = HistoricallyHashableView<C2, <W as ReplaceContext<C2>>::Target>;
+
+    async fn with_context(
+        &mut self,
+        ctx: impl FnOnce(&Self::Context) -> C2 + Clone,
+    ) -> Self::Target {
+        HistoricallyHashableView {
+            _phantom: PhantomData,
+            stored_hash: self.stored_hash,
+            hash: self.hash,
+            inner: self.inner.with_context(ctx).await,
+        }
+    }
+}
+
+impl<W> View for HistoricallyHashableView<W::Context, W>
+where
+    W: View,
+{
+    const NUM_INIT_KEYS: usize = 1 + W::NUM_INIT_KEYS;
+
+    type Context = W::Context;
+
+    fn context(&self) -> &Self::Context {
+        self.inner.context()
+    }
+
+    fn pre_load(context: &Self::Context) -> Result<Vec<Vec<u8>>, ViewError> {
+        let mut v = vec![context.base_key().base_tag(KeyTag::Hash as u8)];
+        let base_key = context.base_key().base_tag(KeyTag::Inner as u8);
+        let context = context.clone_with_base_key(base_key);
+        v.extend(W::pre_load(&context)?);
+        Ok(v)
+    }
+
+    fn post_load(context: Self::Context, values: &[Option<Vec<u8>>]) -> Result<Self, ViewError> {
+        let hash = from_bytes_option(values.first().ok_or(ViewError::PostLoadValuesError)?)?;
+        let base_key = context.base_key().base_tag(KeyTag::Inner as u8);
+        let context = context.clone_with_base_key(base_key);
+        let inner = W::post_load(
+            context,
+            values.get(1..).ok_or(ViewError::PostLoadValuesError)?,
+        )?;
+        Ok(Self {
+            _phantom: PhantomData,
+            stored_hash: hash,
+            hash,
+            inner,
+        })
+    }
+
+    async fn load(context: Self::Context) -> Result<Self, ViewError> {
+        let keys = Self::pre_load(&context)?;
+        let values = context.store().read_multi_values_bytes(keys).await?;
+        Self::post_load(context, &values)
+    }
```

Comment:
(I'm surprised this isn't the default implementation. Isn't it the same for almost all views?)